sycl/doc/extensions/proposed/sycl_ext_oneapi_malloc_shared_context.asciidoc

```diff
@@ -0,0 +1,116 @@
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 6 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Overview
+
+This extension adds new USM allocation methods for `shared` allocations. These new
+methods provide a way to allocate memory without specifying an initial `device`.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_MALLOC_SHARED_CONTEXT` to one of the values defined in 
+the table below.  Applications can test for the existence of this macro to 
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== API Additions
+
+This extension adds the following new member function to SYCL in the 
+`sycl::ext::oneapi` namespace.
+
+```
+namespace sycl {
+  namespace ext {
+    namespace oneapi {
+      void* malloc_shared(size_t numBytes, 
+                          const context& syclContext,
+                          const property_list& propList = {});
+      void* aligned_alloc_shared(size_t alignment, size_t numBytes,
+                                 const context& syclContext,
+                                 const property_list& propList = {});
+
+      template <typename T>
+      T* malloc_shared<T>(size_t count, 
+                          const context& syclContext,
+                          const property_list& propList = {});
+
+      template <typename T>
+      T* aligned_alloc_shared(size_t alignment, size_t count,
+                              const context& syclContext,
+                              const property_list& propList = {});
+    } // namespace oneapi
+  } // namespace ext
+} // namespace sycl 
+
+```
+
+This APIs perform the same functions as defined in 
+https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_shared_allocation_functions[Section 4.8.3.4: Shared allocation functions]. The resulting shared allocations are accessible to every device
+contained in the context. Multiple devices may concurrently access this memory
+only if every device in the context supports the aspect 
+`usm_atomic_shared_allocations`.
```

Comment:
I think we need to have consistent rules about accessibility for these new functions and the existing USM "shared" allocation functions in the SYCL 2020 spec.  For example:

```
void *p1 = sycl::malloc_shared(N, dev, ctxt);
void *p2 = sycl::ext::oneapi::malloc_shared(N, ctxt);
```

1. The memory at `p1` is accessible on `dev` because it was allocated for that device.
2. The memory at `p1` is accessible on other devices in `ctxt` if [some set of rules].
3. The memory at `p2` is accessible on devices in `ctxt` if [some set of rules].

The "[some set of rules]" must be the same in points (1) and (2) in my opinion.

The SYCL spec could be clearer about these rules, but I think the best interpretation is either like this:

> This memory is accessible on some device _D_ if that device is in the memory's allocation context and if that device has the aspect `aspect::usm_atomic_shared_allocations`.

or like this:

> This memory is accessible on some device _D_ if that device is either in the memory's allocation context or it is a descendent device of some other device in that context and if device _D_ has the aspect `aspect::usm_atomic_shared_allocations`.

(Note that these are both different from what you have.)

I'd like to use this extension proposal to agree on this wording if possible, and then I will create a PR clarifying the core SYCL spec.